### PR TITLE
omcproxy: Rework omcproxy.init to make it work again

### PR DESCRIPTION
**Pull request by howl**
Actual omcproxy,init file doesn't bring up omcproxy at boot. This version is based in igmpproxy.init but adapted to omcproxy config file structure.

Changes from the old init are:
Set up network triggers in service_triggers.
Use ubus calls to get network device from interface name.
Use ubus calls to get interface status.
Specify proper family and proto options in firewall rules.

This fixes https://bugs.openwrt.org/index.php?do=details&task_id=1972 where omcproxy will not start up at boot because triggers aren't set properly in that situation and the network interface is not up when omcproxy init file starts. This init version set ups triggers even when no interface is up for the ones defined in omcproxy's config so when interface is ready it brings up.

 Signed-off-by: David Santamaría Rogado <howl.nsp@gmail.com>

### CODE REVIEW

**Comment by dedeckeh** [comment]
The PR has formal issues, no SoB  and no clear git commit description what is changed in the patch. Please have a look at https://openwrt.org/submitting-patches

**Comment by howl** [comment]
My bad, that doesn't edit the commit message, there is a way to edit a commit message?

**Comment by bobafetthotmail** [comment]
The github web interface can't do it, you need a local git folder, see the submitting patches article for details on setting it up.

Then do a 
**git commit --amend**
in your local git, it will open a text editor to change the commit message and then 
**git push -f**
to update the github PR

**Comment by howl** [comment]
@bobafetthotmail , I have practically no idea to use git in command line but I could do it with your help. I only had one problem by editing another commit but could guess that was a branch selection problem.

@dedeckeh had edit now, also with more changes like the number of characters per line and name_of_the_package: in the subject. If there is something more I have done wrong please don't hesitate to point me.I want to add something more to the init file like config file triggers but as is now works similar that the original one but not failing at boot.

File: package/network/services/omcproxy/files/omcproxy.init [hunk 1]
```diff
@@ -1,62 +1,88 @@
...
-	config_get downlink $1 downlink
+	local proxy scope uplink downlinks device up
+
 	config_get scope $1 scope
+	config_get uplink $1 uplink
+	config_get downlinks $1 downlink
+
+	[ -n "$uplink" ] || return
+
+	local status="$(ubus -S call "network.interface.$uplink" status)"
```
Comment (dedeckeh) [inline package/network/services/omcproxy/files/omcproxy.init]: network.sh (/lib/functions/network.sh) has wrapper functions which obsoletes the need to do ubus network calls in scripts

File: package/network/services/omcproxy/files/omcproxy.init [hunk 2]
```diff
@@ -1,62 +1,88 @@
...
+	[ -n "$status" ] || return
+
+	json_load "$status"
+	json_get_var device l3_device
```
Comment (dedeckeh) [inline package/network/services/omcproxy/files/omcproxy.init]: Retrieve the l3_device by using network_get_device

File: package/network/services/omcproxy/files/omcproxy.init [hunk 3]
```diff
@@ -1,62 +1,88 @@
...
+	json_get_var up up
```
Comment (dedeckeh) [inline package/network/services/omcproxy/files/omcproxy.init]: You can check if an interface is up by using the network_is_up

File: package/network/services/omcproxy/files/omcproxy.init [hunk 4]
```diff
@@ -1,62 +1,88 @@
...
+
+	[ -n "$device" -a "$up" = "1" ] || {
+		procd_append_param error "$uplink is not up"
+		return;
+	}
+
+	proxy="$device"
+
+	for downlink in $downlinks; do
+		local status="$(ubus -S call "network.interface.$downlink" status)"
```
Comment (dedeckeh) [inline package/network/services/omcproxy/files/omcproxy.init]: As mentioned above there's no need to use ubus network calls directly if you use the network.sh wrapper functions

File: package/network/services/omcproxy/files/omcproxy.init [hunk 5]
```diff
@@ -1,62 +1,88 @@
...
+		[ -n "$status" ] || return
 
-	proxy=""
+		json_load "$status"
+		json_get_var device l3_device
```
Comment (dedeckeh) [inline package/network/services/omcproxy/files/omcproxy.init]: Use network_get_device() to retrieve the layer3 device

File: package/network/services/omcproxy/files/omcproxy.init [hunk 6]
```diff
@@ -1,62 +1,88 @@
...
+		json_get_var up up
```
Comment (dedeckeh) [inline package/network/services/omcproxy/files/omcproxy.init]: Use network_is_up to check if an interface is up

File: package/network/services/omcproxy/files/omcproxy.init [hunk 7]
```diff
@@ -1,62 +1,88 @@
...
+omcproxy_add_network_triggers() {
+	local uplink downlinks
+
 	config_get uplink $1 uplink
-	config_get downlink $1 downlink
+	config_get downlinks $1 downlink
 
-	for network in $uplink $downlink; do
-		procd_add_interface_trigger "interface.*" $network /etc/init.d/omcproxy restart
+	procd_add_interface_trigger "interface.*" $uplink /etc/init.d/omcproxy reload
```
Comment (dedeckeh) [inline package/network/services/omcproxy/files/omcproxy.init]: I prefer to keep the procd_add_interface_trigger for the uplink in the for loop as before

File: package/network/services/omcproxy/files/omcproxy.init [hunk 8]
```diff
@@ -104,14 +130,13 @@ omcproxy_add_firewall() {
 
 service_triggers() {
 	procd_add_reload_trigger "omcproxy"
+	config_foreach omcproxy_add_network_triggers proxy
 }
 
 start_service() {
-	include /lib/functions
```
Comment (dedeckeh) [inline package/network/services/omcproxy/files/omcproxy.init]: This should not be removed if you make use of network.sh wrapper functions

**Review by dedeckeh — CHANGES_REQUESTED** [review]

**Comment by howl** [comment]
@dedeckeh I did it with network functions the first time but changed it because testing with ash could see that network status was cached till calling a function to refresh the statuses returned by network functions. I should have asked before to not work two times.

I have pushed another commit, now network functions are used. There is no need to check status as I could see that getting device name results null if interface doesn't exist or is not up.

Also uplink merged in downlinks for as was in the original one.

I will see to change also igmpproxy as it use ubus calls too, but I need to get more git skills.

As before, anything that need to be changed just ask for it.

I have tested the new version and still works as expected.

File: package/network/services/omcproxy/files/omcproxy.init (hunk 1 above)
Comment (howl) [reply package/network/services/omcproxy/files/omcproxy.init]: Done.

**Comment by dedeckeh** [comment]
@howl The PR has now 2 commits which makes the PR impossible to merge.
The first commit has the original changes while the second commit fixes my remarks which should have been in the original commit.
You should really try to stay away from the github user interface and rather fork the project to a public repo and clone the repo to a computer.
Instructions and guidance is provided in https://openwrt.org/submitting-patches#working_with_github

**Comment by howl** [comment]
@dedeckeh tested your version and it works right, setting repeated interfaces and using PROCD_DEBUG=1 shows that your code works right avoiding repeated triggers. But, I could see that firewall rules anyways get repeated, but at first sight I don't know a way to do the same as you did for the triggers with the firewall rules without setting up another more global variable.

It could be commit as is now because it's better than the actual init script and it could be improved over the time.

The only little thing I think could be changed is in omcproxy_add_proxy(), moving "config_get downlinks $1 downlink" just before the downlinks for and "config_get scope $1 scope" just before "[ -n "$scope" ] && proxy="$proxy,scope=$scope"" to avoid getting configuration variables from proxies where no upstream is defined.

Also in omcproxy_add_network_triggers() a "[ -n "$uplink" ] || return" should be added just after getting upstream config variable, as triggers for no valid proxies are added.

If you think that are interesting changes I could make a patch.

**Comment by dedeckeh** [comment]
I will close this PR as a reworked commit has been pushed (https://git.openwrt.org/?p=openwrt/openwrt.git;a=commit;h=df8f8bad082ea5280c19c9091c099b28b7e2fc67). I had to undo the change in omcproxy_add_proxy to set proxy=$device as it would break the check [ -n "$proxy" ] || return 0 if no downlink devices are found

**Pull request closed without merging**